torch/_inductor/runtime/triton_helpers.py

```diff
@@ -614,3 +614,49 @@ def select_one(x, mask, dim, keep_dims=False):
     ix = x.to(idtype, bitcast=True)
     iy = tl.sum(ix * mask, dim, keep_dims=keep_dims)
     return iy.to(x.dtype, bitcast=True)
+
+
+@triton.jit
+def gpu_barrier(
+    sem, total: tl.constexpr, first_barrier: tl.constexpr = tl.constexpr(False)
+):
+    """
+    Wait for all other thread blocks to reach this barrier before returning.
+
+    if *sem <= total:   we are in count_up mode waiting for all threads to enter
+    if *sem >= total*2: we are in count_down mode waiting for all threads to leave
+
+    Args:
+        sem: an int32 semaphores, zero initialized
+        total: how many other blocks are running
+        first_barrier: True if we can skip waiting for exit from last barrier.  Can set either on the first call,
+                       or if one ping-pongs between two semaphores.
+    """
+    # ensure stores before this are visible
```

Comment:
we should adapt grid.syncI() implementation from cooperative groups with inline_asm_elementwise, it makes only a single thread from the block do the loads, and avoids atomic ops in the countdown phase (as soon as the thread sees the expected value it's free to go, there's no need to actually atomically reduct counter all the way down). The `bar_has_flipped` still allows one to call sync multiple times without resetting the semaphore:
```
_CG_STATIC_QUALIFIER bool bar_has_flipped(unsigned int old_arrive, unsigned int current_arrive) {
    return (((old_arrive ^ current_arrive) & 0x80000000) != 0);
}

_CG_STATIC_QUALIFIER void sync_grids(unsigned int expected, volatile barrier_t *arrived) {
    bool cta_master = (threadIdx.x + threadIdx.y + threadIdx.z == 0);
    bool gpu_master = (blockIdx.x + blockIdx.y + blockIdx.z == 0);

    __barrier_sync(0);

    if (cta_master) {
        unsigned int oldArrive;
        unsigned int nb = 1;
        if (gpu_master) {
            nb = 0x80000000 - (expected - 1);
        }

        // Barrier update with release; polling with acquire
        asm volatile("atom.add.release.gpu.u32 %0,[%1],%2;" : "=r"(oldArrive) : "l"((unsigned int*)arrived), "r"(nb) : "memory");

        unsigned int current_arrive;
        do {
            asm volatile("ld.acquire.gpu.u32 %0,[%1];" : "=r"(current_arrive) : "l"((unsigned int *)arrived) : "memory");
        } while (!bar_has_flipped(oldArrive, current_arrive));
    }

    __barrier_sync(0);
}
```
This is the code from cooperative_groups cuda header.

Reply:
Something like this
```
@triton.jit
def get_tid():
    return tl.inline_asm_elementwise(
        """
        mov.u32 $0, %tid.x;
        mov.u32 $1, %tid.y;
        mov.u32 $2, %tid.z;
        """,
        "=r,=r,=r",
        [],
        dtype=(tl.uint32, tl.uint32, tl.uint32),
        is_pure=True,
        pack=1,
    )

@triton.jit
def cta_master():
    tx, ty, tz = get_tid()
    return tx + ty + tz == 0

@triton.jit
def gpu_master():
    return tl.program_id(0) + tl.program_id(1) + tl.program_id(2) == 0

@triton.jit
def grid_barrier(sem):
    tl.inline_asm_elementwise(
        "bar.sync 0;", "=r", [], dtype=tl.int32, is_pure=False, pack=1
    )
    if cta_master():
        expected = tl.num_programs(0) * tl.num_programs(1) * tl.num_programs(2)
        if gpu_master():
            nb = (0x80000000 - (expected - 1)).to(tl.int32)
        else:
            nb = 1
        old_arrive = tl.inline_asm_elementwise("atom.add.release.gpu.u32 $0,[$1],$2;",
        "=r,l,r", [sem,nb], dtype=tl.int32, is_pure=False, pack=1)
        bar_flipped = False
        while not bar_flipped:
            current_arrive = tl.inline_asm_elementwise("ld.acquire.gpu.u32 $0,[$1];",
            "=r,l", [sem], dtype=tl.int32, is_pure=False, pack=1)
            bar_flipped = ((old_arrive ^ current_arrive) & 0x80000000) != 0
    tl.inline_asm_elementwise(
        "bar.sync 0;", "=r", [], dtype=tl.int32, is_pure=False, pack=1
    )
```

Reply:
The implementation sets sem to 0x80000000 after first sync call, and to 0 after the second etc, and `bar_flipped = ((old_arrive ^ current_arrive) & 0x80000000) != 0` correctly detects flip condition in both cases. 
It syncs all the blocks but can be adapted to only sync on `tl.num_programs(0)` but yeah it's non-portable.

Reply:
@ngimel this is my attempt at a more portable version of the same algorithm:

```py
@triton.jit
def x_grid_barrier(sem):
    tl.debug_barrier()

    one_i32 = 1
    one_u32 = one_i32.to(tl.uint32)
    expected = tl.num_programs(0).to(tl.uint32)
    if tl.program_id(0) == 0:
        nb = (0x80000000 - (expected - one_u32))
    else:
        nb = one_u32

    old_arrive = tl.atomic_add(sem, nb, sem="release")

    bar_flipped = False
    while not bar_flipped:
        # want a `ld.acquire.gpu.u32 $0,[$1];` but Triton doesn't have it
        # current_arrive = tl.atomic_add(sem, 0, sem="acquire")
        current_arrive = tl.load(sem, volatile=True)  # is missing .acquire
        bar_flipped = ((old_arrive ^ current_arrive) & 0x80000000) != 0

    tl.debug_barrier()
```

I think this generates nearly the code we want, the problem is generating the
```
ld.acquire.gpu.u32
```
we could either do:
```
current_arrive = tl.load(sem, volatile=True)
```
which is missing the `.acquire`, or
```
current_arrive = tl.atomic_add(sem, 0, sem="acquire")
```
which seems wasteful, but should have the same semantics as the `ld.acquire`.


Do you fully understand why the `acquire` is needed?  Is it so the entire function acts a memory fence?

Reply:
Yeah `acquire` is needed to ensure memory consistency, subsequent reads cannot be seen as happening before the acquire, which is what we want here.  `__threadfence` would have worked but is a bigger hammer, in fact pre-volta version of grid sync has threadfences before atomic add and after polling. `atomic_add` in triton has an advantage that it's predicated to happen only on 1 thread, regular load would have all threads poll the location, but I agree it's still wasteful.

Reply:
btw `volatile` should work too, it flushes all the caches so should have the needed effect. Except it's happening on all threads w/o predicate, so is kinda wasteful (and volatile is bigger hammer than needed)